### PR TITLE
Create secret

### PR DESCRIPTION
**Pull request by LimKianAn**
for issue #93

### CODE REVIEW

File: api/v1/postgres_types.go [hunk 1]
```diff
@@ -215,6 +217,50 @@ func (p *Postgres) ToKey() *types.NamespacedName {
 	}
 }
 
+const SecretNamespace = "pgaas"
+
+func (p *Postgres) ToSecret(src *corev1.SecretList) (*corev1.Secret, error) {
+	new := &corev1.Secret{}
```
Comment (majst01) [inline api/v1/postgres_types.go:223]: new is a reserved word and irritates reading

File: api/v1/postgres_types.go [hunk 2]
```diff
@@ -215,6 +217,50 @@ func (p *Postgres) ToKey() *types.NamespacedName {
...
+			},
+		),
+	}
+}
+
+func (p *Postgres) ToSecretName() string {
+	return p.ToPeripheralResourceName() + "-passwords"
+}
+
+func (p *Postgres) ToSecretNamesapcedName() *types.NamespacedName {
```
Comment (majst01) [inline api/v1/postgres_types.go:257]: typo

File: api/v1/postgres_types.go [hunk 3]
```diff
@@ -215,6 +217,50 @@ func (p *Postgres) ToKey() *types.NamespacedName {
...
+			map[string]string{
+				"application":  "spilo",
+				"cluster-name": p.ToPeripheralResourceName(),
+				"team":         p.generateTeamID(),
+			},
+		),
+	}
+}
+
+func (p *Postgres) ToSecretName() string {
```
Comment (LimKianAn) [inline api/v1/postgres_types.go:253]: `ToUserPasswordsSecretName()`

File: controllers/postgres_controller.go [hunk 4]
```diff
@@ -162,6 +162,12 @@ func (r *PostgresReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, fmt.Errorf("unable to create or update corresponding CRD ClusterwideNetworkPolicy: %W", err)
 	}
 
+	if instance.Spec.SecretName == "" {
```
Comment (LimKianAn) [inline controllers/postgres_controller.go:165]: Remove the check. Always update the secret, replacing the `Data` block.

**Comment by LimKianAn** [comment]
Decision: `Secret`s which reside in control-plane cluster are in the same namespace as `Postgres`. For example: pgaas

Suggestion: 
1. The name of `Secret`s should be the name of `Postgres` plus a suffix.
2. `secretName` should not be in the spec. of `Postgres`. It's a direct derivation of unknown spec. of `Postgres`.

File: api/v1/postgres_types.go [hunk 5]
```diff
@@ -280,6 +283,42 @@ func (p *Postgres) ToPeripheralResourceName() string {
 	return p.generateTeamID() + "-" + p.generateDatabaseName()
 }
 
+// ToUserPasswordSecret returns the secret containing user password pairs
+func (p *Postgres) ToUserPasswordsSecret(src *corev1.SecretList, scheme *runtime.Scheme) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	secret.Namespace = p.Namespace
+	// todo: Consider `p.Name + "-passwords", so the`
```
Comment (eberlep) [inline api/v1/postgres_types.go:290]: You're right, I think `p.Name` would be better here: We create the secret in the control plane cluster, so we should use the name of the resource in the control plane cluster, not the name iof the resource n the service cluster.

File: controllers/status_controller.go [hunk 6]
```diff
@@ -107,16 +109,61 @@ func (r *StatusReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+	result, err := controllerutil.CreateOrUpdate(ctx, r.Control, fetched, func() error {
+		fetched.Data = secret.Data
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create or update the secret containing user password pairs: %w", err)
+	}
+	log.Info("secret created or updated", "operation result", result)
+
+	if result == controllerutil.OperationResultCreated {
```
Comment (eberlep) [inline controllers/status_controller.go:151]: You're right here as well: The name will probably not change and could be derived automativally (by convention). Hence, we would not need the field secretName in the spec.

What we should do, however, is export the String-Suffix we use to generate the secret name (currently `-password`) so it the same string will be used in the `cloudctl`/cloud-api

File: controllers/status_controller.go [hunk 7]
```diff
@@ -107,16 +109,61 @@ func (r *StatusReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+	}
+
 	return ctrl.Result{}, nil
 }
 
 // SetupWithManager registers this controller for reconciliation of Postgresql resources
 func (r *StatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&zalando.Postgresql{}).
+		// Owns(&corev1.Secret{}). // todo: better the watching on secrets that weren't generated by this controller
```
Comment (eberlep) [inline controllers/status_controller.go:129]: Not used, please remove.

**Review by eberlep — CHANGES_REQUESTED** [review]

File: api/v1/postgres_types.go (hunk 5 above)
Comment (majst01) [reply api/v1/postgres_types.go:290]: cloud-api must be able to detect the name of the secret, currently the backup-secret is stored under:
uuid-backup
Where uuid is the full blown uuid / name of the postgres instance

Comment (LimKianAn) [reply api/v1/postgres_types.go:290]: `secretName` is `postgresName-passowrds` now.

File: controllers/status_controller.go (hunk 6 above)
Comment (LimKianAn) [reply controllers/status_controller.go:151]: @majst01
We added an exported func `ToUserPasswordsSecretName` to `Postgres`. Can we remove `secretName` from the spec.?

**Review by eberlep — APPROVED** [review]

**Pull request merged**